Fix #7029: ALR hashes are added to source when DONOTROCKETOPTIMIZE is set to true

Miraeld opened the pull request:
# Description

Fixes #7029
When `DONOTROCKETOPTIMIZE` is set to true, WP Rocket should completely disable all optimization features. However, Performance Hints (LRC/LCP) hashes were still being added to pages and warmup processes were still running, which could impact site performance and user experience.

This fix ensures that when `DONOTROCKETOPTIMIZE` is enabled:
- No Performance Hints hashes are added to pages
- No warmup processes are triggered for LRC/LCP
- The beacon injection remains disabled (as previously implemented in #7022)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue).


## Detailed scenario

### What was tested

**Automated Tests:**
- Unit tests for LazyRenderContent Context `is_allowed()` method with `DONOTROCKETOPTIMIZE` scenarios
- Integration tests for WarmUp Controller methods (`warm_up()` and `warm_up_home()`) with `DONOTROCKETOPTIMIZE` scenarios
- All existing Performance Hints tests continue to pass

**Manual Scenarios:**
- **Before fix**: With define('DONOTROCKETOPTIMIZE', true); in wp-config.php, visiting pages still showed Performance Hints hashes in the HTML
- **After fix**: With define('DONOTROCKETOPTIMIZE', true); in wp-config.php, no Performance Hints hashes are added to pages
- Warmup behavior: No warmup jobs are queued when DONOTROCKETOPTIMIZE is enabled

### How to test

**Test Steps:**

- Add to `wp-config.php`: `define('DONOTROCKETOPTIMIZE', true);`
- Visit any page in incognito mode
- Check page source - no Performance Hints hashes should be present
- Check that no warmup jobs are triggered in the background

## Technical description

### Documentation

This pull request introduces a new condition to respect the `DONOTROCKETOPTIMIZE` constant, ensuring that certain optimizations are bypassed when this constant is set. Corresponding changes include updates to logic in the codebase and the addition of test cases to validate this behavior.

### Core Logic Updates:
* Added a check for the `DONOTROCKETOPTIMIZE` constant in the `is_allowed` method of `WarmUp\Controller` to prevent optimizations when the constant is set. (`inc/Engine/Common/PerformanceHints/WarmUp/Controller.php`, [inc/Engine/Common/PerformanceHints/WarmUp/Controller.phpR70](diffhunk://#diff-6c58efcf5a5487d78bd3c705d9afb577c66f93c78e3b5bd89d123c8928733629R70))
* Updated the `is_allowed` method in `LazyRenderContent\Context` to include a similar check for the `DONOTROCKETOPTIMIZE` constant. (`inc/Engine/Optimization/LazyRenderContent/Context/Context.php`, [inc/Engine/Optimization/LazyRenderContent/Context/Context.phpR20-R23](diffhunk://#diff-d50d2b45f846b62ca2d432ad6e1f68e8c0dcd18c7bddda2edd57275dc3f72623R20-R23))

### Test Enhancements:
* Added new test cases to verify the behavior when `DONOTROCKETOPTIMIZE` is set, ensuring optimizations are skipped in `WarmUp` and `LazyRenderContent` contexts. (`tests/Fixtures/inc/Engine/Common/PerformanceHints/WarmUp/Subscriber/warmUp.php`, [[1]](diffhunk://#diff-2b82d71d243c77c2d7ebc6e0b9300cc7f2cb84012a53dda77a14d7f461c031dbR55-R68); `tests/Fixtures/inc/Engine/Common/PerformanceHints/WarmUp/Subscriber/warmUpHome.php`, [[2]](diffhunk://#diff-6b8f3b70c392996bac5e620416da02fa5194a3b32131abf9ae02fbc034a966eaR43-R53); `tests/Fixtures/inc/Engine/Optimization/LazyRenderContent/Context/Context/isAllowed.php`, [[3]](diffhunk://#diff-5e9b5b2b727db94cda639a1727aaad1ea91d534d61c353eef068d264a595d579R7-R31)
* Updated integration tests to mock the `DONOTROCKETOPTIMIZE` constant and validate its impact on behavior. (`tests/Integration/inc/Engine/Common/PerformanceHints/WarmUp/Subscriber/warmUp.php`, [[1]](diffhunk://#diff-8b05cdb08109b53e9c5038e703ccadb17644025705fe44e1510e0a35b3253098L36-R40); `tests/Integration/inc/Engine/Common/PerformanceHints/WarmUp/Subscriber/warmUpHome.php`, [[2]](diffhunk://#diff-d19752a588566cec07ae5722d1d9965c956ee4f86bc357815fccf173b1b3b0d9L36-R43)
* Adjusted unit tests to include the `DONOTROCKETOPTIMIZE` constant in the configuration and verify its influence on the `isAllowed` logic. (`tests/Unit/inc/Engine/Optimization/LazyRenderContent/Context/Context/isAllowed.php`, [tests/Unit/inc/Engine/Optimization/LazyRenderContent/Context/Context/isAllowed.phpR27](diffhunk://#diff-16ce53ad575872bcd4f82f076def96a1bf330272ccb25e2fe6f0ff2b0ded006aR27))
### New dependencies

None

### Risks

None

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

Miraeld committed:
Add support for DONOTROCKETOPTIMIZE constant in performance hints and lazy render context

Miraeld committed:
fix tests

jeawhanlee commented on inc/Engine/Optimization/LazyRenderContent/Context/Context.php:
```diff
@@ -17,6 +18,10 @@ public function is_allowed( array $data = [] ): bool {
 			return false;
 		}
 
+		if ( rocket_get_constant( 'DONOTROCKETOPTIMIZE' ) && ! Utils::is_saas_visit() ) {
```
We don't need a check with `Utils::is_saas_visit()`. We already do that [here](https://github.com/wp-media/wp-rocket/blob/00ed622a2f6ac34448de53136ce5df82171dcb8f/inc/Engine/Common/PerformanceHints/Frontend/Subscriber.php#L70) before making [rocket_performance_hints_buffer](https://github.com/wp-media/wp-rocket/blob/00ed622a2f6ac34448de53136ce5df82171dcb8f/inc/Engine/Common/PerformanceHints/Frontend/Subscriber.php#L96) available which is then used [here](https://github.com/wp-media/wp-rocket/blob/4b9193bc96b46fd1e8a5db6f3a9bf4126a155eb5/inc/Engine/Optimization/LazyRenderContent/Frontend/Subscriber.php#L33) for only adding hashes on sass visits.

Miraeld replied:
Updated

Miraeld committed:
Simplify condition for DONOTROCKETOPTIMIZE in lazy render context

Miraeld committed:
Merge branch 'develop' into fix/7029-alr-hashes-added-to-source

Miraeld committed:
Merge branch 'develop' into fix/7029-alr-hashes-added-to-source

Miraeld committed:
Fix lrc status in beacon data for performance hints test

wordpressfan committed:
use nowdoc instead of heredoc

wordpressfan committed:
Merge branch 'develop' into fix/7029-alr-hashes-added-to-source

Miraeld committed:
Refactor lazy render optimization checks to move constant checks

jeawhanlee commented on inc/Engine/Optimization/LazyRenderContent/Frontend/Controller.php:
```diff
@@ -181,7 +185,7 @@ private function get_exclusions(): array {
 	 */
 	public function add_custom_data( array $data ): array {
-		$data['status']['lrc'] = $this->context->is_allowed();
+		$data['status']['lrc'] = ! rocket_get_constant( 'DONOTROCKETOPTIMIZE' ) && $this->context->is_allowed();
```
We don't need the extra check here, we already do that before injecting the beacon [here](https://github.com/wp-media/wp-rocket/blob/00ed622a2f6ac34448de53136ce5df82171dcb8f/inc/Engine/Common/PerformanceHints/Frontend/Processor.php#L99)
```suggestion
		$data['status']['lrc'] = $this->context->is_allowed();
```

jeawhanlee committed:
Removed unnecessary check

jeawhanlee committed:
Updated test data

wordpressfan commented:
When we visit any page that can't be cached (which is the case with sending the warmup request) we set the constant `DONOTROCKETOPTIMIZE` to be true here:
https://github.com/wp-media/wp-rocket/blob/8b0e79150b6b32c9cd23883a6077212e9d943fba/inc/classes/Buffer/class-cache.php#L82-L86

So this condition will fail here:
https://github.com/wp-media/wp-rocket/blob/ba4d3e885c4aa520752f99eb5a3e66c9e38095fb/inc/Engine/Common/PerformanceHints/WarmUp/Controller.php#L83-L85

Miraeld committed:
Merge branch 'develop' into fix/7029-alr-hashes-added-to-source

Miraeld committed:
Fix: Remove DONOTROCKETOPTIMIZE check from WarmUp to prevent circular logic

The WarmUp Controller should not check DONOTROCKETOPTIMIZE because:
- WP Rocket internally sets this constant during warmup requests
- Checking it would break the warmup feature entirely

The LazyRenderContent checks remain in place to properly handle
user-defined DONOTROCKETOPTIMIZE in wp-config.php (fixes #7029).

This addresses the issue identified by @wordpressfan in PR #7515.

Miraeld commented:
## Fixed Circular Logic Issue

Thanks to @wordpressfan for identifying the problem! I've updated the PR to fix the circular logic issue.

### What Changed

**Removed:**
- ❌ `DONOTROCKETOPTIMIZE` check from WarmUp Controller
- ❌ Related test scenarios for WarmUp

**Kept (these are correct):**
- ✅ `DONOTROCKETOPTIMIZE` check in LazyRenderContent Context
- ✅ `DONOTROCKETOPTIMIZE` check in LazyRenderContent Frontend Controller
- ✅ All LazyRenderContent tests

### Why This Is Correct

**The Original Issue (#7029) - STILL FIXED:**
When a developer adds `define('DONOTROCKETOPTIMIZE', true);` to wp-config.php, Performance Hints hashes are now properly prevented from being added to pages. This is handled by the LazyRenderContent checks.

**The Circular Logic Problem - NOW FIXED:**
WP Rocket internally sets `DONOTROCKETOPTIMIZE` during warmup requests (in `Buffer/Cache.php` [lines 82-86](https://github.com/wp-media/wp-rocket/blob/8b0e79150b6b32c9cd23883a6077212e9d943fba/inc/classes/Buffer/class-cache.php#L82-L86)). If we checked this constant in the WarmUp Controller, it would:
1. Try to run warmup
2. WP Rocket sets `DONOTROCKETOPTIMIZE` during the warmup request
3. WarmUp sees the constant and stops
4. Warmup never completes ❌

**The Solution:**
- WarmUp should NOT check `DONOTROCKETOPTIMIZE` (it's an internal process)
- LazyRenderContent SHOULD check `DONOTROCKETOPTIMIZE` (it handles user-facing pages)

### Files Changed in Latest Commit
```
inc/Engine/Common/PerformanceHints/WarmUp/Controller.php        (removed check)
tests/Fixtures/.../WarmUp/Subscriber/warmUp.php                 (removed test scenario)
tests/Fixtures/.../WarmUp/Subscriber/warmUpHome.php             (removed test scenario)
tests/Integration/.../WarmUp/Subscriber/warmUp.php              (removed mocking)
tests/Integration/.../WarmUp/Subscriber/warmUpHome.php          (removed mocking)
```

Ready for re-review! 🚀

hanna-meda commented:
Thank you @Miraeld for this PR.

When testing the warmup of priority elements with `DONOTROCKETOPTIMIZE` set to true in wp-config.php, URLs are still being fetched on a fresh install and when permalinks are changed.

Steps to reproduce:
1. Set DONOTROCKETOPTIMIZE in wp-config.php
2. Fresh install and activate the PR zip
3. Check the LCP, LRC tables

Current behavior: URLs are fetched and added to the priority elements tables.

Miraeld committed:
Fix: Add DONOTROCKETOPTIMIZE check to add_hashes method

When DONOTROCKETOPTIMIZE is set, LazyRenderContent hashes were still
being added during SaaS/Performance Hints visits because the add_hashes()
method didn't check the constant.

Changes:
- Add DONOTROCKETOPTIMIZE check to add_hashes() method
- Add test coverage for DONOTROCKETOPTIMIZE in add_hashes test
- Fix test bug in add_hashes_when_allowed that was overriding filter

This ensures DONOTROCKETOPTIMIZE is respected in ALL scenarios:
- Regular site visits (via add_hashes_when_allowed)
- Performance Hints collection visits (via add_hashes)

Fixes the failing manual test scenario reported in PR comments

Miraeld commented:
## Fixed the Failing Manual Test Scenario

I've identified and fixed the issue reported in the PR comments. The problem was that `DONOTROCKETOPTIMIZE` wasn't being respected during **SaaS/Performance Hints collection visits**.

### Root Cause

The `add_hashes()` method (called on `rocket_performance_hints_buffer` during SaaS visits) did NOT check `DONOTROCKETOPTIMIZE`, while `add_hashes_when_allowed()` (called on `rocket_buffer` for regular visits) did.

**Failing scenario:**
1. User sets `define('DONOTROCKETOPTIMIZE', true);` in wp-config.php
2. SaaS makes a Performance Hints collection visit
3. ❌ Hashes were still being added via `add_hashes()` 

### Changes Made

**Core Fix:**
- Added `DONOTROCKETOPTIMIZE` check to `add_hashes()` method in [Controller.php](https://github.com/wp-media/wp-rocket/blob/b6956bb6a/inc/Engine/Optimization/LazyRenderContent/Frontend/Controller.php#L143-L146)

**Test Coverage:**
- Added new test case `shouldNotAddHashesWhenDONOTROCKETOPTIMIZE` for the SaaS visit scenario
- Fixed existing test bug in `add_hashes_when_allowed.php` where `__return_true` was overriding the filter

### Verification

✅ All PerformanceHints tests pass: **87 tests, 98 assertions**  
✅ New test verifies DONOTROCKETOPTIMIZE is respected during SaaS visits  
✅ Commit c1c37fc remains correct (was fixing a different WarmUp issue)

Now `DONOTROCKETOPTIMIZE` is properly respected in **ALL scenarios**:
- Regular site visits (via `add_hashes_when_allowed`)
- Performance Hints collection visits (via `add_hashes`)

The manual test scenario should now pass! 🎉

Miraeld committed:
Fix: Add DONOTROCKETOPTIMIZE check to WarmUp Controller

Prevents warmup from populating LCP/LRC database tables when DONOTROCKETOPTIMIZE is set.
This ensures no URLs are fetched and no database entries are created during fresh install or permalink changes when the constant is enabled.

hanna-meda commented:
@Miraeld, checking after latest changes on PR, now the reverse has happened, no URLs are ever fetched/warmed-up: 

1. `DONOTROCKETOPTIMIZE` NOT in the wp-config.php file.
2. Fresh install / change permalink / change theme.
3. Check LCP, LRC tables. Optionally, if WPR debug on, also search inside wp-rocket-debug.log.html file for "Add to queue request arguments" logs. 

Current behavior: No URLs are fetched. Inside the "wp-rocket-debug.log.html" file there are only 2 logs of home URL.
